Added tag log option to json-logger

bonczj opened the pull request:
Fixes #19803
Updated the json-logger to utilize the common log option
'tag' that can define container/image information to include
as part of logging.

When the 'tag' log option is not included, there is no change
to the log content via the json-logger. When the 'tag' log option
is included, the tag will be parsed as a template and the result
will be stored within each log entry as the attribute 'tag'.

Signed-off-by: bonczj <josh.bonczkowski@gmail.com>

**- What I did**
Enabled the --log-opt tag= for the json-logger to allow operators to also log variables from https://docs.docker.com/engine/admin/logging/log_tags/.

**Note**: This is a continuation from #34088. The new PR is due to my botched rebase on the original. I'm not sure the best way to clean up it, so instead I created a new branch from my changes plus the code review change and rebased a bit more correctly.

**- How I did it**
Enabled the 'tag' option on json-logger. Utilized the existing loggerutils.ParseLogTag method to parse the tag template while providing an empty default template. Updated the jsonlog-marshaling routines in pkg/jsonlog/ to allow writing the tag string if it was specified.

**- How to verify it**
Use --log-driver json-logger --log-opt tag="[tag template]" either through 'docker run' or through dockerd.

**- Description for the changelog**
Added tag log option to json-logger

<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->


**- A picture of a cute animal (not mandatory but encouraged)**
![cute animal](http://cuteanimalpicturesandvideos.com/wp-content/uploads/lizard.jpg)

bonczj commented:
@jnovack - because of my botched rebase, I found it easier to just move to a new PR. I have included the change based on your code review. Please look it over with the change.

bonczj commented:
The one failure on powerpc I'm not sure about. It does not appear to be related to the change. I'm not sure if it is a build environment or something else.

`out: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded`

Looks a bit like an environmental timeout.

dnephin commented on daemon/logger/jsonfilelog/jsonfilelog_test.go:
```diff
@@ -54,6 +54,51 @@ func TestJSONFileLogger(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "docker-logger-")
+	if err != nil {
+		t.Fatal(err)
+	}
```
```go
require.NoError(t, err)
```

dnephin commented on pkg/jsonlog/jsonlog_marshalling.go:
```diff
@@ -101,6 +101,15 @@ func (mj *JSONLog) MarshalJSONBuf(buf *bytes.Buffer) error {
+		}
+		buf.WriteString(`"tag":`)
+		ffjsonWriteJSONString(buf, mj.Tag)
+	}
```
Did you use `ffjson` to generate this code? According to the comment on this file that is how it was done originally.

bonczj replied:
No, I skipped to where it broke and fixed it thereby missing the comment at the top that I did not anticipate. I'll work on applying the ffjson generation code as well as the patch that is in the comment section at the top of the file.

bonczj replied:
I followed the process documented in the file. I ran the command 'go get -u github.com/pquerna/ffjson' and then 'ffjson pkg/jsonlog/jsonlog.go'. Upon opening the new file (jsonlog_ffjson.go), the contents look nothing like the original source. 

I suspect the ffjson tool has been updated, but it is substantially different. I see an unmarshal function, but not a marshal function. I have the new output from ffjson over at https://gist.github.com/bonczj/6f683afbc888c8d517601d79c6a7acb7.

dnephin replied:
Ok, I looked into this and opened #34946 to fix the comment. Don't worry about generating the code from ffjson, what you have will work for now.

dnephin commented on daemon/logger/jsonfilelog/jsonfilelog_test.go:
```diff
@@ -54,6 +54,51 @@ func TestJSONFileLogger(t *testing.T) {
+`
+
+	if string(res) != expected {
+		t.Fatalf("Wrong log content: %q, expected %q", res, expected)
```
```go
assert.Equal(t, expected, string(res)
```

dnephin commented on integration-cli/docker_cli_logs_test.go:
```diff
@@ -305,3 +305,14 @@ func (s *DockerSuite) TestLogsWithDetails(c *check.C) {
 	c.Assert(details[1], checker.Equals, "foo=bar")
 }
+
+func (s *DockerSuite) TestLogsWithTag(c *check.C) {
```
`integration-cli` is deprecated, we should not be adding any more tests to this suite (#34407 is pending to document this).

Unit tests should be sufficient for testing this change.

bonczj replied:
I see it in the docs now. I'll have this backed out.

vdemeester commented:
ping @bonczj

bonczj commented:
Hey, I'm just getting back from a vacation. It'll be a few more days before I catch up on my daily work to reply.

Evertvz commented:
Hello @bonczj 

Would you have some time to check the review? That'd be great thanks.

Best regards,
Evert

bonczj commented:
Thank you @Evertvz for drawing my attention back here.

I am looking at the changes over the next day or two. Most of them are straight forward. The more complicated part is getting it all to build again :)

bonczj commented:
I have applied the changes to the tests as @dnephin requested. I will work on the ffjson issue after hearing an update to my problem above.

The test failures for experimental, windowsRS1 and z appear to be unrelated to the changes.

thaJeztah commented:
I restarted CI

@dnephin could you have a look? https://github.com/moby/moby/pull/34248#discussion_r137646066

dnephin commented:
I'll look into it today. I do wonder if this is an unnecessary optimization and we should just remove it.

dnephin commented on pkg/jsonlog/jsonlog.go:
```diff
@@ -17,6 +17,8 @@ type JSONLog struct {
 	// Attrs is the list of extra attributes provided by the user
 	Attrs map[string]string `json:"attrs,omitempty"`
+	// Tags requested the operator
+	Tag string `json:"tag,omitempty"`
```
This should be set to `""` in `Reset()`

bonczj replied:
Tag is now initialized to an empty string in Reset().

dnephin commented on daemon/logger/jsonfilelog/jsonfilelog_test.go:
```diff
@@ -54,6 +56,49 @@ func TestJSONFileLogger(t *testing.T) {
+	tmp, err := ioutil.TempDir("", "docker-logger-")
+	if err != nil {
+		require.NoError(t, err)
+	}
```
You don't need to check for `err != nil` first, that's what `require.NoError()` does. Just remove the if condition.

bonczj replied:
Done.

dnephin commented on daemon/logger/jsonfilelog/jsonfilelog_test.go:
```diff
@@ -54,6 +56,49 @@ func TestJSONFileLogger(t *testing.T) {
+	})
+	if err != nil {
+		t.Fatal(err)
+	}
```
`require.NoError(t, err)`

same with all the lines below that check for `err != nil`

bonczj replied:
Done for the new method I added. I did not correct the rest of the file.

dnephin commented on pkg/jsonlog/jsonlogbytes.go:
```diff
@@ -38,6 +39,15 @@ func (mj *JSONLogs) MarshalJSONBuf(buf *bytes.Buffer) error {
+		}
+		buf.WriteString(`"tag":`)
+		ffjsonWriteJSONString(buf, mj.Tag)
+	}
```
Please add a case that uses `tag` to the list of testcases for this function

bonczj replied:
Done. Added an extra log line with a tag to the existing unit test for it.

dnephin commented on pkg/jsonlog/jsonlog_marshalling.go:
```diff
@@ -101,6 +101,15 @@ func (mj *JSONLog) MarshalJSONBuf(buf *bytes.Buffer) error {
 		buf.WriteString(`"stream":`)
 		ffjsonWriteJSONString(buf, mj.Stream)
 	}
+	if len(mj.Tag) > 0 {
```
Please add a case that uses `tag` to the list of testcases for this function

bonczj replied:
Done. Added an extra log line with a tag to the existing unit test for it.

dnephin commented on pkg/jsonlog/jsonlogbytes_test.go:
```diff
@@ -23,6 +23,7 @@ func TestJSONLogsMarshalJSONBuf(t *testing.T) {
 		{Log: []byte{0x7F}}:            `^{\"log\":\"\x7f\",\"time\":}$`,
 		// with raw attributes
 		{Log: []byte("A log line"), RawAttrs: []byte(`{"hello":"world","value":1234}`)}: `^{\"log\":\"A log line\",\"attrs\":{\"hello\":\"world\",\"value\":1234},\"time\":}$`,
+		{Log: []byte("A log line with tag"), Tag: "test-tag"}:                           `^{\"log\":\"A log line with tag\",\"tag\":\"test-tag\",\"time\":}$`,
```
I fixed the other cases in my PR (#34946). Please remove the `}$` from the end of the regex here. We shouldn't expect invalid json, and this will be fixed in #34946

dnephin reviewed (approved):
LGTM, just one small test comment

bonczj commented:
@dnephin I just pushed the one small test fix you requested. 

I want you to be aware that similar regex are found in pkg/jsonlog/jsonlog_marshalling_test.go.

dnephin commented:
Thanks, yes I think I fixed the other ones in #34946

bonczj commented:
Dang it @dnephin ! Your change set went in first, now I have conflicts! ;) Do you want to pick up some merge work? hah.

I'm working on merging your changes. Nice to see some of the older work removed to simplify the system. It will be easier for the next person.

bonczj committed:
Added tag log option to json-logger

Fixes #19803
Updated the json-logger to utilize the common log option
'tag' that can define container/image information to include
as part of logging.

When the 'tag' log option is not included, there is no change
to the log content via the json-logger. When the 'tag' log option
is included, the tag will be parsed as a template and the result
will be stored within each log entry as the attribute 'tag'.

Update: Removing test added to integration_cli as those have been deprecated.
Update: Using proper test calls (require and assert) in jsonfilelog_test.go based on review.
Update: Added new unit test configs for logs with tag. Updated unit test error checking.
Update: Cleanup check in jsonlogbytes_test.go to match pending changes in PR #34946.
Update: Merging to correct conflicts from PR #34946.

Signed-off-by: bonczj <josh.bonczkowski@gmail.com>

bonczj commented:
@dnephin I have (hopefully) corrected the merge conflicts between our PRs. I suspect you will want to review the latest changes now.

dnephin commented:
sorry about that

dnephin reviewed (approved):
LGTM

thaJeztah commented:
My concern with this PR is still that the log-files written by the json-file logger are considered an implementation detail, and only for use by the daemon (to return the log-entries when calling `docker logs`). Adding features/additional information to this file that are not used by the daemon itself effectively makes the file format now an API (whereas currently the format is not defined, and could be modified at any given point)

cpuguy83 commented on daemon/logger/jsonfilelog/jsonlog/jsonlogbytes.go:
```diff
@@ -12,6 +12,7 @@ type JSONLogs struct {
 	Log     []byte    `json:"log,omitempty"`
 	Stream  string    `json:"stream,omitempty"`
 	Created time.Time `json:"time"`
+	Tag     string    `json:"tag,omitempty"`
```
What's the reasoning behind making a new field instead of using `RawAttrs`?

bonczj replied:
Mostly for clarity and (slightly?) easier parsing later.

thaJeztah commented:
Oh, looks like this needs a rebase now :cry:

@cpuguy83 do you prefer https://github.com/moby/moby/pull/34248#discussion_r146294849 to be changed to use the `RawAttrs`?

tbatchelli commented:
@thaJeztah so if the json-file logs are deemed internal, is there a plan to have an external file logger that others can use? Logging to file should be a possibility, no?

cpuguy83 commented:
There is a plugin API for log drivers. See https://github.com/cpuguy83/docker-log-driver-test/blob/master/driver.go as an example that just makes the built-in json-file log driver a plugin.

cpuguy83 commented:
@thaJeztah I think RawAttrs is probably what we want here.

bonczj commented:
I want to point out that I will not have the opportunity to come back to this any time soon. In Oct I moved to a new company and my role and work load have shifted considerably.

yongtang commented:
Carried this PR to #35949, also use RawAttr instead. Please take a look.
